### PR TITLE
ssh/tailssh: add logic for matching against AcceptEnv patterns

### PR DESCRIPTION
**Pull request by mpminardi**
Add logic for parsing and matching against our planned format for AcceptEnv values. Namely, this supports direct matches against string values and matching where * and ? are treated as wildcard characters which match against an arbitrary number of characters and a single character respectively.

Actually using this logic in non-test code will come in subsequent changes.

Updates https://github.com/tailscale/corp/issues/22775

### CODE REVIEW

File: ssh/tailssh/accept_env_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		{pattern: "*", target: "EXAMPLE_ENV", match: true},
+		{pattern: "?", target: "A", match: true},
+		{pattern: "?*", target: "EXAMPLE_2", match: true},
+		{pattern: "??", target: "CC", match: true},
+		{pattern: "?*?", target: "ABCDEFG", match: true},
+		{pattern: "*?*", target: "ABCDEFG", match: true},
+		{pattern: "***", target: "123456", match: true},
+		{pattern: "**?TEST", target: "_TEST", match: true},
+		{pattern: "TEST**?", target: "TEST_", match: true},
+		{pattern: "TEST_*", target: "TEST_A", match: true},
```
Comment (oxtoacart) [inline ssh/tailssh/accept_env_test.go:29]: It's technically covered in the below test, but it might be nice to also add a test case where the target is longer than a pattern terminated in `*`.

File: ssh/tailssh/accept_env.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+				}
+
+				patternIdx++
+			}
+			matchingAsterisk = true
+			continue
+		}
+
+		if matchingAsterisk {
+			// We want to find increase the index in the target string
```
Comment (oxtoacart) [inline ssh/tailssh/accept_env.go:93]: "We want to find increase" -> "We want to increase"

File: ssh/tailssh/accept_env.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	}
+
+	return false
+}
+
+// matchPattern returns true if the pattern matches against the target string.
+// Patterns may include * and ? wildcard characters which match against an
+// arbitrary number of characters or a single character respectively.
+func matchPattern(pattern string, target string) bool {
```
Comment (oxtoacart) [inline ssh/tailssh/accept_env.go:58]: Consider renaming to `matchAcceptEnvPattern` since this is a fairly large package.

File: ssh/tailssh/accept_env_test.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+package tailssh
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestMatchPattern(t *testing.T) {
```
Comment (oxtoacart) [inline ssh/tailssh/accept_env_test.go:13]: Consider renaming to `TestMatchAcceptEnvPattern`.

File: ssh/tailssh/accept_env_test.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		{pattern: "EXAMPLE_*_ENV", target: "EXAMPLE_aBcd2231---_ENV", match: true},
+		{pattern: "COMPLICA?ED_PATTERN*", target: "COMPLICATED_PATTERN_REST", match: true},
+		{pattern: "COMPLICAT???ED_PATT??ERN", target: "COMPLICAT123ED_PATTggERN", match: true},
+		{pattern: "DIRECT_MATCH", target: "DIRECT_MATCH", match: true},
+
+		// Negative cases
+		{pattern: "?", target: "123", match: false},
+		{pattern: "??", target: "123", match: false},
+		{pattern: "?*?", target: "A", match: false},
+		{pattern: "*?*", target: "C", match: false},
```
Comment (oxtoacart) [inline ssh/tailssh/accept_env_test.go:40]: Consider including these two tests

```
{pattern: "*", target: "", match: false},
{pattern: "?*", target: "A", match: false},
```

Comment (oxtoacart) [reply ssh/tailssh/accept_env_test.go:40]: I would expect the below to pass, but it fails

```
{pattern: "*?", target: "A", match: false},
```

Comment (mpminardi) [reply ssh/tailssh/accept_env_test.go:40]: Discussed offline, the current behaviour mirrors openSSH's logic for handling patterns so will stay as-is.

Namely:
- `?` matches one character
- `*` matches zero, one, or many

Comment (mpminardi) [reply ssh/tailssh/accept_env_test.go:40]: Additional test cases will be brought over from https://github.com/openssh/openssh-portable/blob/46d1fb16b20e971b9ac15e86a3d3e350b49c9ad6/regress/unittests/match/tests.c (openSSH's test suite for their implementation) to catch any remaining edge cases / to ensure compatibility.

File: ssh/tailssh/accept_env_test.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestMatchAcceptEnvPattern(t *testing.T) {
+	testCases := []struct {
+		pattern string
+		target  string
+		match   bool
+	}{
+		// Positive cases
```
Comment (oxtoacart) [inline ssh/tailssh/accept_env_test.go:19]: Nit - it might be easier to follow the tests if positive and negative test cases for the same pattern were grouped together, as opposed to grouping all positive test cases together.

**Review by oxtoacart — APPROVED** [review]
Just one nit about test case ordering, otherwise LGTM.

File: ssh/tailssh/accept_env_test.go (hunk 6 above)
Comment (mpminardi) [reply ssh/tailssh/accept_env_test.go:19]: Oh fair, like have them grouped like:

```go
{pattern: "TEST_*", target: "TEST_A_LONG_ENVIRONMENT_VARIABLE_NAME", match: true},
{pattern: "TEST_*", target: "TEST", match: false},

{pattern: "??", target: "CC", match: true},
{pattern: "??", target: "123", match: false},
```

you mean?
